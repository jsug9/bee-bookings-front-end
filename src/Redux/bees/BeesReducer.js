/* eslint-disable no-param-reassign */
import axios from 'axios';
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const GET_BEES = 'Bees/GET_BEES';
const GET_BEE_DETAILS = 'Bees/GET_BEE_DETAILS';
const ADD_BEE = 'Bees/ADD_BEE';
const DELETE_BEE = 'Bees/DELETE_BEE';

const getBeesEndpoint = 'https://bee-store.herokuapp.com/api/v1/items';

const initialState = {
  allBees: [],
  beeDetails: {},
  isLoading: false,
};

export const getAllBees = createAsyncThunk(GET_BEES, async () => {
  const { data } = await axios.get(getBeesEndpoint);
  return data;
});

export const getBeeDetails = createAsyncThunk(GET_BEE_DETAILS, async (id) => {
  const { data } = await axios.get(`${getBeesEndpoint}/${id}`);
  return data;
});

export const addBee = createAsyncThunk(ADD_BEE, async (bee) => {
  axios.post(getBeesEndpoint, bee);
  const { data } = await axios.get(getBeesEndpoint);
  return data;
});

export const deleteBee = createAsyncThunk(DELETE_BEE, async (beeId) => {
  axios.delete(`${getBeesEndpoint}/${beeId}`);
  const { data } = await axios.get(getBeesEndpoint);
  return data;
});

const beesSlice = createSlice({
  name: 'Bees',
  initialState,
  reducers: {},
  extraReducers: {
    [getAllBees.fulfilled]: (state, action) => {
      state.allBees = action.payload.sort((a, b) => a.id - b.id);
      state.isLoading = false;
    },
    [getAllBees.rejected]: (state) => {
      state.allBees = [];
      state.isLoading = false;
    },
    [getAllBees.pending]: (state) => {
      state.allBees = [];
      state.isLoading = true;
    },
    [getBeeDetails.fulfilled]: (state, action) => {
      state.beeDetails = action.payload;
      state.isLoading = false;
    },
    [getBeeDetails.rejected]: (state) => {
      state.beeDetails = {};
      state.isLoading = false;
    },
    [getBeeDetails.pending]: (state) => {
      state.beeDetails = {};
      state.isLoading = true;
    },
    [addBee.fulfilled]: (state, action) => {
      state.allBees = action.payload;
      state.isLoading = false;
    },
    [addBee.rejected]: (state) => {
      state.allBees = [];
      state.isLoading = false;
    },
    [addBee.pending]: (state) => {
      state.allBees = [];
      state.isLoading = true;
    },
    [deleteBee.fulfilled]: (state, action) => {
      state.allBees = action.payload;
      state.isLoading = false;
    },
    [deleteBee.rejected]: (state) => {
      state.allBees = [];
      state.isLoading = false;
    },
    [deleteBee.pending]: (state) => {
      state.allBees = [];
      state.isLoading = true;
    },
  },
});

export default beesSlice.reducer;
