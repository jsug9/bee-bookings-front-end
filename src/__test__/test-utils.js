/* eslint-disable react/prop-types */
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import setupStore from '../Redux/testStore';

function renderWithProviders(
  ui,
  {
    preloadedState,
    store = setupStore(preloadedState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Router><Provider store={store}>{children}</Provider></Router>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';

export default renderWithProviders;
