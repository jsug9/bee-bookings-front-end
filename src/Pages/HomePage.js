import { useSelector, useDispatch } from 'react-redux';
import React, { useEffect } from 'react';
// import { getBees } from '../Redux/bees/BeesReducer';
import BeesList from '../Components/BeesList';
import '../Styles/Homepage.scss';
import { getAllBees } from '../Redux/bees/BeesReducer';

const HomePage = () => {
  const dispatch = useDispatch();
  const bees = useSelector((state) => state.bees.allBees);

  useEffect(() => {
    dispatch(getAllBees());
  }, []);

  return (
    <section>
      <div className="homepage-title">
        <h1 className="title">Our Collection of Bees</h1>
        <p className="description">Please select a Bee</p>
      </div>
      <BeesList bees={bees} />
    </section>
  );
};

export default HomePage;
