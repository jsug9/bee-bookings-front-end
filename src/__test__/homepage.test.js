// import axios from 'axios';
import { act } from 'react-dom/test-utils';
import HomePage from '../Pages/HomePage';
import renderWithProviders, { screen } from './test-utils';
import { getAllBees } from '../Redux/bees/BeesReducer';
import server from '../mswMocks/server';
import realStore from '../Redux/configureStore';

// Establish API mocking before all tests.
beforeAll(() => server.listen());
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());
// Clean up after the tests are finished.
afterAll(() => server.close());

it('Homepage renders to the page', () => {
  const tree = renderWithProviders(
    <HomePage />,
  );
  expect(tree).toMatchSnapshot();
});

it('HomePage text is shown in the page', async () => {
  renderWithProviders(
    <HomePage />,
  );

  expect(screen.getByText('Our Collection of Bees')).toBeInTheDocument();
});

it('Bee cards render in the page', async () => {
  const store = realStore;
  await act(async () => {
    renderWithProviders(
      <HomePage />, { store },
    );
    await store.dispatch(getAllBees());
  });

  expect(store.getState().bees.allBees.length).toBe(10);
});
