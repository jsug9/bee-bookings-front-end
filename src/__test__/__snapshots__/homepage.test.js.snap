// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage renders to the page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section>
        <div
          class="homepage-title"
        >
          <h1
            class="title"
          >
            Our Collection of Bees
          </h1>
          <p
            class="description"
          >
            Please select a Bee
          </p>
        </div>
        <div
          class="react-multi-carousel-list  carousel-container"
          dir="ltr"
        >
          <ul
            class="react-multi-carousel-track "
            style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
          />
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section>
      <div
        class="homepage-title"
      >
        <h1
          class="title"
        >
          Our Collection of Bees
        </h1>
        <p
          class="description"
        >
          Please select a Bee
        </p>
      </div>
      <div
        class="react-multi-carousel-list  carousel-container"
        dir="ltr"
      >
        <ul
          class="react-multi-carousel-track "
          style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
        />
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
